Source/WebCore/platform/graphics/cg/ImageDecoderCG.cpp

```diff
@@ -589,18 +585,73 @@ PlatformImagePtr ImageDecoderCG::createFrameImageAtIndex(size_t index, Subsampli
     return maskedImage ? maskedImage : image;
 }
 
+String ImageDecoderCG::decodeUTI(const SharedBuffer& data) const
+{
+    auto uti = String(CGImageSourceGetType(m_nativeDecoder.get()));
+    if (uti != "public.heif"_s && uti != "public.heic"_s)
+        return uti;
+    
+    constexpr auto bytesToUnsigned = [](const char bytes[4]) -> unsigned {
+        return bytes[0] | bytes[1] << 8 | bytes[2] << 16 | bytes[3] << 24;
+    };
+
+    static constexpr unsigned ftypSignature = bytesToUnsigned("ftyp");
+    static constexpr unsigned avifBrand = bytesToUnsigned("avif");
+    static constexpr unsigned avisBrand = bytesToUnsigned("avis");
+
+    // HEIF/AVIF files start with an ftyp box
+    //
+    //  aligned(8) class GeneralTypeBox(code) extends Box(code) {
+    //      unsigned int(32)    major_brand;
+    //      unsigned int(32)    minor_version;
+    //      unsigned int(32)    compatible_brands[];    // to end of the box
+    //  }
+    //
+    // A box starts with 4 bytes for the size (including the length bytes).
+    // An image is considered an AVIF if "avif" or "avis" is present in its
+    // major_brand or in one of the compatible_brands.
+
+    const unsigned* ftypBox = reinterpret_cast<const unsigned*>(data.data());
+
+    // The size is in little endian.
+    unsigned sizeInBytes = flipBytes(ftypBox[0]);
+    if (sizeInBytes % 4 || sizeInBytes <= 8)
+        return uti;
+
+    if (ftypBox[1] != ftypSignature)
+        return uti;
+
+    unsigned sizeInUnsigned = sizeInBytes / sizeof(unsigned);
+
+    for (unsigned i = 2; i < sizeInUnsigned; ++i) {
```

Comment:
Should we impose some reasonable limit so we don't walk through a potentially huge box size?

Reply:
Done. I limit it to check only 100 compatible_brands.